ouroboros-consensus/src/Ouroboros/Consensus/Node.hs

```diff
@@ -464,34 +522,54 @@ forkBlockProduction IS{..} =
   New network layer
 -------------------------------------------------------------------------------}
 
-data NetworkRequires m up blk hdr = NetworkRequires {
-      -- | Start a chain sync client that communicates with the given upstream
-      -- node.
-      nrChainSyncClient     :: up -> ChainSyncClient hdr (Point hdr) m Void
-
-      -- | Start a chain sync server.
-    , nrChainSyncServer     :: ChainSyncServer hdr (Point hdr) m ()
-
-      -- | Start a block fetch client that communicates with the given
-      -- upstream node.
-    , nrBlockFetchClient    :: BlockFetchClient hdr blk m ()
-
-      -- | Start a block fetch server server.
-    , nrBlockFetchServer    :: BlockFetchServer blk m ()
-
-      -- | The fetch client registry, used by the block fetch client.
-    , nrFetchClientRegistry :: FetchClientRegistry up hdr blk m
+-- | Interface which 'ouroboros-network' provides.  We put it in a record,
+-- since at the moment 'ouroboros-network' only provides this interface in 'IO'.
+-- This will change in the future and this record will not be needed.
+--
+-- XXX: Maybe it is not needed!
+--
+data NetworkInterface m up bytes = NetworkInterface {
+      -- | Like 'Ouroboros.Network.NodeToNode.nodeToNodeConnectTo'
+      --
+      niConnectTo
+        :: forall a.
+           Versions NodeToNodeVersion
+                    DictVersion
+                    (MuxApplication InitiatorApp NodeToNodeProtocols m bytes a)
+        -> up
+        -> m ()
+
+      -- | Like 'Ouroboros.Network.NodeToNode.withServerNodeToNode'
+      --
+    , niWithServerNode
+        :: forall t.
+           (forall vData. DictVersion vData -> vData -> vData -> Accept)
+        -> Versions NodeToNodeVersion
+                    DictVersion
+                    (AnyMuxResponderApp NodeToNodeProtocols m bytes)
+        -> (Async m () -> m t)
+        -> m t
     }
 
--- | Required by the network layer to initiate comms to a new node
-data NodeComms m ps e bytes = NodeComms {
-      -- | Codec used for the protocol
-      ncCodec    :: Codec ps e m bytes
-
-      -- | Construct a channel to the node
+data NetworkRequires m up blk hdr bytes a b = NetworkRequires {
+      -- | Mux application which runs with upstream peers.
+      --
+      nrMuxInitiatorApp     :: up
```

Comment:
I would expect this `Versions` stuff to be encapsulated. This is specific to a muxed communication channel.

Reply:
I agree, if we abstract over it it will be easier to do different things: in this form handshake negotation is forced upon the node which I am not happy with.